### ISSUE TITLE
Postgres sync corrupts blocks from 98493

### ISSUE THREAD

**Issue by mvanhalen** [deso-protocol/core#211]
When doing a fresh sync with the latest version of core. Fresh node, fresh DB blocks are not processed correctly from block 98493. 

Seems like some cache mismatch. Will check it out.

See error:

`I0205 11:21:58.422185       1 server.go:1147] Server._handleBlock: Received block ( 98492 / 101262 ) from Peer [ Remote Address: 34.123.41.111:17000 PeerID=1 ]
I0205 11:21:58.661121       1 server.go:1147] Server._handleBlock: Received block ( 98493 / 101262 ) from Peer [ Remote Address: 34.123.41.111:17000 PeerID=1 ]
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map

goroutine 116 [running]:
github.com/go-pg/pg/v10.(*Tx).RunInTransaction.func1()
        /root/go/pkg/mod/github.com/go-pg/pg/v10@v10.10.0/tx.go:91 +0x86
panic({0x10a30a0, 0x16fdd80})
        /usr/lib/go/src/runtime/panic.go:1038 +0x215
github.com/deso-protocol/core/lib.(*Postgres).flushPosts(0xa33af3, 0xc0bac47f90, 0xc0009e1680)
        /deso/src/core/lib/postgres.go:1462 +0x94a
github.com/deso-protocol/core/lib.(*Postgres).FlushView.func1(0x0)
        /deso/src/core/lib/postgres.go:1303 +0x7d
github.com/go-pg/pg/v10.(*Tx).RunInTransaction(0xc0bb0334c0, {0x172c7e8, 0xc00003a290}, 0xc0b3b1d638)
        /root/go/pkg/mod/github.com/go-pg/pg/v10@v10.10.0/tx.go:95 +0xa5
github.com/go-pg/pg/v10.(*baseDB).RunInTransaction(0xc00003a290, {0x172c7e8, 0xc00003a290}, 0xc0b3b1d688)
        /root/go/pkg/mod/github.com/go-pg/pg/v10@v10.10.0/tx.go:74 +0x51
github.com/deso-protocol/core/lib.(*Postgres).FlushView(0xb1df40, 0xc000010040)
        /deso/src/core/lib/postgres.go:1296 +0x57
github.com/deso-protocol/core/lib.(*UtxoView).FlushToDb(0xc0009e1680)
        /deso/src/core/lib/block_view_flush.go:16 +0x34
        /deso/src/core/lib/blockchain.go:1942 +0x1946
github.com/deso-protocol/core/lib.(*Server)._handleBlock(0xc0002103c0, 0xc0003eed80, 0xc0bcf05110)
        /deso/src/core/lib/server.go:1193 +0x3e5
github.com/deso-protocol/core/lib.(*Server)._handlePeerMessages(0xc0002103c0, 0xc1128a2fc0)
        /deso/src/core/lib/server.go:1494 +0x18d
github.com/deso-protocol/core/lib.(*Server).messageHandler(0xc0002103c0)
        /deso/src/core/lib/server.go:1534 +0x19c
created by github.com/deso-protocol/core/lib.(*Server).Start
        /deso/src/core/lib/server.go:1710 +0xb1`

**Comment by tijno** [comment]
Confirmed on my end too. Also mentioned by kanshi on discord.

**Comment by StarGeezerPhil** [comment]
Thank you so much for the alert on this - I'll pause my sync (~80k) in the hope we can fix this without _another_ resync.

**Comment by lazynina** [comment]
@mvanhalen - thanks for flagging this. I'm out of pocket this weekend, but will investigate first thing on Monday.

**Comment by mvanhalen** [comment]
> @mvanhalen - thanks for flagging this. I'm out of pocket this weekend, but will investigate first thing on Monday.

Thanks for the update. Will do a little debugging. Those GO conversion things are pretty weird for me. c# background.

**Comment by lazynina** [comment]
@mvanhalen I see the issue. We need to initialize the make with a make(map[string]uint64) before we start filling it in.

**Comment by mvanhalen** [comment]
> @mvanhalen I see the issue. We need to initialize the make with a make(map[string]uint64) before we start filling it in.

Thanks will check that out.

**Comment by mvanhalen** [comment]
Working with @tijno on this @lazynina .

Added 
`
AdditionalNFTRoyaltiesToCoinsBasisPoints: make(map[string]uint64),
AdditionalNFTRoyaltiesToCreatorsBasisPoints: make(map[string]uint64),
`

below extra data on https://github.com/deso-protocol/core/blob/main/lib/postgres.go#L1447

It compiled and running that for a sync. Will know the result tomorrow.

**Comment by diamondhands0** [comment]
Hey all, just an update here.

@lazynina put out a fix in this PR:
* https://github.com/deso-protocol/core/pull/212

We are now re-running a full sync from scratch to make sure there aren't any more surprises, but feel free to give it a shot yourselves in the meantime.

I also wanted to take a moment to apologize for the pain you guys felt on this one. Across the core team, we have treated Postgres as a secondary data store, not used in consensus or our production nodes, and not yet integrated into our extremely thorough test suite.

The reason for this is that we have some fear that Postgres will not scale well as DeSo usage increases. And so, while it's useful for apps getting off the ground, we have yet to make a go or no-go decision with regard to whether we want to offer full long-term support for it.

The above being said, given that serious apps like NFTz are now relying on it, I think we are leaning toward working around Postgres's limitations. In particular, we have made a few breakthroughs with a new syncing mechanism we call Hypersync that, if applied to our Postgres path, should allow it to scale reasonably well.

All this to say: We're trying really hard to support Postgres at the level that we support all core code, and we appreciate your patience with us as we work out the kinks. I know how useful it is to have a SQL interface over the DeSo data, and we're doing everything we can to keep this a viable path for you guys long-term.

**Comment by tijno** [comment]
great thanks @diamondhands0 im at block 58k of resync and just rebuild backend with this change added. Another 8 hours or so and ill know if it got to the end.

**Comment by mvanhalen** [comment]
Thanks @diamondhands0 and @lazynina , very good to hear that. We have no issues switching to MySql or something else if needed. We use an ORM layer just for that. But we do need data for NFTz. We have a hybrid model (postgres and normal node) so 60% is operational still for buying and selling, but the stats and other functionality depend on postgres now.

Looking forward to the new innovations and let us know where we can help, debug or whateva. 

Hypersync sound like music. A general backup and restore kind of service might work as well. Like a daily general backup we can use as start or something.

Running a sync as well at block 71K now (updated a previous running one).

**Comment by diamondhands0** [comment]
My resync just hit another issue FYI. We're digging into it. At this point it may be faster to just bite the bullet and hook Postgres into our test suite.

I0206 13:38:53.910761  784421 server.go:1147] Server._handleBlock: Received block ( 98599 / 101808 ) from Peer [ Remote Address: 34.123.41.
111:17000 PeerID=177 ]                                                                                                                     
E0206 13:38:53.927366  784421 server.go:1141] Server._handleBlock: Encountered an error processing block <Header: < 98599, 0000000000004d02
262295ffe7f63fe770ce4442d095927113cd25a6938d11f1, 1 >, Signer Key: BC1YLh768bVj2R3QpSiduxcvn7ipxF3L3XHsabZYtCGtsinUnNrZvNN>. Disconnecting 
from peer [ Remote Address: 34.123.41.111:17000 PeerID=177 ]: Error while processing block: : ConnectBlock: : ConnectTransaction: : _connec
tAcceptNFTBid: _helpConnectNFTSold: Profile missing for additional coin royalty for pub key: BC1YLjSCiP67Zq3v7fwc613Akd8hvQrs2ChuPPqRFTfFQG
dmtomNR1J tBCKYZSjZN4yvxqraWvY3dQZ2SjJjNdxHSyK7WLzU34TDYjL26XoRq

**Comment by mvanhalen** [comment]
Thanks for the update and headsup @diamondhands0 sounds like a good idea,

**Comment by StarGeezerPhil** [comment]
Thank you for the regular updates on this now @diamondhands0

**Comment by diamondhands0** [comment]
Just an update here: Still hunting down the issue. I believe it was due to the encoding of the royalty maps, which I just fixed in this commit:
* https://github.com/deso-protocol/core/pull/212/commits/4809196253d78a6dee8493d8b18a7e6e39c1b2c6

My sync runs pretty quickly, so my hope is that this works, and we can defer integrating Postgres into our test suite, which is a bigger effort than I expected.

**Comment by tijno** [comment]
k great ive just rebuild with those changes and sync was at 49k so lets see if that manages to remain ok.

**Comment by mvanhalen** [comment]
Cool, just rebuild and syncing from 94500 so will let all know.

**Comment by mvanhalen** [comment]
Just synced past 98599 till 98714. Got a new error. @diamondhands0  @tijno  Which is progress.

` Server._handleBlock: Encountered an error processing block <Header: < 98714, 0000000000005428375f6ee0666ae8bde2cc3                          432da490d53f8588341c542b4a6, 1 >, Signer Key: BC1YLh768bVj2R3QpSiduxcvn7ipxF3L3XHsabZYtCGtsinUnNrZvNN>. Disconnecting from peer [ Remote Address: 35.232.92.5:17000 PeerID=9 ]: 

**Error while processing block: : ProcessBlock: Problem upserting block and transactions: InsertTransactionTx: Unimplemented txn type MESSAGING_GROUP**
`


Seems group messaging transactions have and issue with Postgres

**Comment by diamondhands0** [comment]
Just hit the same error myself. Should be able to fix it tonight. Luckily this one shouldn't require a resync after it's fixed.

**Comment by mvanhalen** [comment]
> Just hit the same error myself. Should be able to fix it tonight. Luckily this one shouldn't require a resync after it's fixed.

Nice sync time. Same network datacenter I guess? Thanks for the continued efforts.

**Comment by tijno** [comment]
thx @diamondhands0 for merging additional fixes - will this generate new backend docker image ? right now building from source but prefer to use backend images if all ok

**Comment by diamondhands0** [comment]
I think it's finally fixed. I just finished syncing a Postgres node, and it made it all the way to the end. This was the final commit that was needed to fix all the issues:
* https://github.com/deso-protocol/core/commit/1b9d4d61499d66f3c3f2911496785aa6b95dd5d5

I cut a new release v2.0.5 so just rebooting your nodes should get you to the end if you had the MessagingGroup issue.

**Please let me know once your syncs complete!**

**Comment by tijno** [comment]
ok syncing again after this change

at 98819 and progressing ok

**Comment by tijno** [comment]
sync complete @diamondhands0

**Comment by mvanhalen** [comment]
> sync complete @diamondhands0

Awesome. Great work @diamondhands0 @tijno  catching up

**Comment by tijno** [comment]
Looks like image for backend has not yet been regenerated ?

stable and latest still 4 days old.

usually its pretty instant.

**Comment by mvanhalen** [comment]
Fully synced now as well!

**Comment by StarGeezerPhil** [comment]
Green with envy here 🤣

@diamondhands0 @lazynina as per @tijno's comment above, if you could update the docker image that'd be awesome 👍

**Comment by tijno** [comment]
@diamondhands0 @maebeam backend images have not been updated yet

**Comment by mvanhalen** [comment]
> I think it's finally fixed. I just finished syncing a Postgres node, and it made it all the way to the end. This was the final commit that was needed to fix all the issues:
> 
> * [1b9d4d6](https://github.com/deso-protocol/core/commit/1b9d4d61499d66f3c3f2911496785aa6b95dd5d5)
> 
> I cut a new release v2.0.5 so just rebooting your nodes should get you to the end if you had the MessagingGroup issue.
> 
> **Please let me know once your syncs complete!**

This resolves the issue and  docker images have been updated
